### ISSUE TITLE
fail to check models, get different output from eran model and onnx model

### ISSUE THREAD

**Issue by JaneHJ** [eth-sri/eran#106]
Hello,

Excellent work on ERAN! I am trying to run ERAN on a new model which only contains 2D convolution and Maxpool. The input is a 32*32*3 image and output is a vector of shape (1, 1, 2). when i check the model, it failed on all test samples and the detail of the results on some sample is shown below. The number for 'eran' is the last layer output of the ERAN model(where the LB and UB are the same with the input), 'model 'is output of onnx model which is trained with Pytorch, 'gt' is for the ground truth. we can see that the error between the onnx model and ground truth is small. which means the onnx model works well. While the error on the ERAN model is big.
Any idea why this happen and how to solve it? Would it be better if I only use the 2D conv and remove the Maxpool? Or is there any tricks on the training of the network if we want it to work well on ERAN?
Any thoughts, pls let me know.

Thanks again for the great work!

```
eran, [22.45658732 21.45113453], 
model, [20.752302 19.964554]
gt, [21 20]
started divergence at layer, 0, outputname, 11, 
difference, [-0.00635895 -0.19411361  0.27611786 ... -0.32006019 -0.37774784
 -0.21209315]

eran, [12.00341192 29.55069391], 
model, [18.597437 27.139414]
gt, [18 28]
started divergence at layer, 0, outputname, 11, 
difference, [ 0.14733622  0.08350691  1.33278507 ... -0.92877057  0.231993
 -0.70928615]

eran, [16.20110469 23.64350543], 
model, [22.032396 25.652857]
gt, [21 26]
started divergence at layer, 0, outputname, 11, 
difference, [ 0.08141201  0.04975749  0.92625538 ...  0.44841941  0.91627241
 -0.20528137]

eran, [12.68772288 30.17839799], 
model, [16.274736 27.341152]
gt, [16 28]
started divergence at layer, 0, outputname, 11, 
difference, [ 0.0599338   0.0242741   0.93594114 ... -0.19657298  0.13709676
 -0.29129541]

eran, [12.54802165 30.27682323], 
model, [23.763256 28.019072]
gt, [23 28]
started divergence at layer, 0, outputname, 11, 
difference, [ 0.06228063  0.00599487  0.93162197 ... -0.0075279   0.1628929
 -0.17152696]

eran, [ 8.785477   22.53118144], 
model, [23.084154 27.283724]
gt, [22 28]
started divergence at layer, 0, outputname, 11, 
difference, [-0.01146841 -0.66223659  0.03704047 ... -0.27320452 -0.12197634
 -0.31861686]

eran, [17.22318447 20.98868645], 
model, [20.915415 25.213446]
gt, [21 25]
started divergence at layer, 0, outputname, 11, 
difference, [ 0.07396364 -0.04313423  0.87729995 ...  0.49402098  0.84290184
 -0.12503251]
```

**Comment by mnmueller** [comment]
Hello @JaneHJ,

Thanks for your interest in ERAN. The training should not affect this behaviour assuming that you set epsilon to a very small value. Typically the output is only 1d, thus having a (pseudo) 3d output might cause unexpected information. Without more information, it is hard to pinpoint the origin of this problem. I would encourage you to check, that the right normalization of the data is provided, and that inputs are loaded in the right dimension order.

Cheers,
Mark

**Comment by JaneHJ** [comment]
Hi @mnmueller,

Thanks very much! Will check the result, test further and come back.

thanks!

**Comment by JaneHJ** [comment]
Hi @mnmueller ,

After I trained a new model and carefully confirmed the dataset and the normalization, IT WORKS! 

Thanks very much!

**Comment by hrishitchaudhuri** [comment]
Hello!

I am at present facing a similar problem. I'm trying to run an ONNX model trained on a custom dataset with custom testing data on DeepPoly. The model is classifying the images correctly (outputs a class between 0 and 2); however, the `eran.analyze_box()` method continuously fails to provide the correct classification (the error is sporadic, and by sheer chance, can sometimes lead to a correct classification). I am attaching some of the error logs generated by the testing framework:

Here, `iisc` is the dataset name, and `iisc_onnx` is the name of the network. Inputs are 3-channel images of 49 x 49 pixels and the output is a single class label. I've also tried normalizing the dataset similar to how CIFAR10 has been normalized (by calculating the means and standard deviations of each channel and adjusting the pixels accordingly), without any change in DeepPoly's behavior.

```
iisc, iisc_onnx, deeppoly, 
eran, [  3.66195583 -10.58648638   1.60119863], 
model, [-49.111183 -49.34993   70.20109 ]
iisc, iisc_onnx, deeppoly, started divergence at layer, 0, outputname, input, 
difference, [  5.00808289  69.25181117   1.79370056 ...   9.20582308 -24.96678855
 -14.55233963]


iisc, iisc_onnx, deeppoly, 
eran, [-2.24988285 -0.23672853 -2.6013685 ], 
model, [-8.896298e+00  9.178391e+00  8.875757e-03]
iisc, iisc_onnx, deeppoly, started divergence at layer, 0, outputname, input, 
difference, [ 30.47691971 -62.6650438   45.92621645 ... -14.17474118  14.53823687
  15.17558348]


iisc, iisc_onnx, deeppoly, 
eran, [-13.80325121 -13.60489883  15.0025049 ], 
model, [-14.585453  44.40837  -21.412405]
iisc, iisc_onnx, deeppoly, started divergence at layer, 0, outputname, input, 
difference, [-20.87886117 -23.65085068  13.7533781  ...  -8.02816828   2.34988719
  -5.18007112]
```

**Comment by GgnDpSngh** [comment]
Hi there,

Eran expects inputs in the format height, width, channel, the most likely cause of this error is passing inputs in the format channel, height, width. Can you check whether you are passing input in the expected format?

Cheers,
Gagandeep Singh

**Comment by hrishitchaudhuri** [comment]
Hello!

Thank you for this, I transposed the inputs and all is working fine now!
